File: rhizome/dht/routing_table.py
"""
Таблица маршрутизации Kademlia (k-бакеты)
"""

import time
from typing import List, Optional

from rhizome.dht.node import Node, NodeID
from rhizome.utils.crypto import compute_distance


class KBucket:
    """k-бакет для хранения узлов на определенном расстоянии"""

    def __init__(self, k: int = 20):
        self.k = k
        self.nodes: List[Node] = []
        self.last_updated = time.time()

    def add_node(self, node: Node) -> bool:
        """
        Добавление узла в бакет

        Returns:
            True если узел добавлен, False если бакет полон
        """
        # Если узел уже есть, перемещаем его в конец (LRU)
        if node in self.nodes:
            self.nodes.remove(node)
            self.nodes.append(node)
            self.last_updated = time.time()
            return True

        # Если есть место, добавляем
        if len(self.nodes) < self.k:
            self.nodes.append(node)
            self.last_updated = time.time()
            return True

        # Бакет полон
        return False

    def remove_node(self, node: Node):
        """Удаление узла из бакета"""
        if node in self.nodes:
            self.nodes.remove(node)
            self.last_updated = time.time()

    def get_nodes(self, limit: Optional[int] = None) -> List[Node]:
        """Получение узлов из бакета"""
        nodes = self.nodes
        if limit:
            nodes = nodes[:limit]
        return nodes

    def is_full(self) -> bool:
        """Проверка, полон ли бакет"""
        return len(self.nodes) >= self.k


class RoutingTable:
    """Таблица маршрутизации Kademlia"""

    def __init__(self, node_id: NodeID, k: int = 20, bucket_count: int = 160):
        self.node_id = node_id
        self.k = k
        self.buckets: List[KBucket] = [KBucket(k) for _ in range(bucket_count)]

    def _get_bucket_index(self, target_id: NodeID) -> int:
        """Получение индекса бакета для целевого ID"""
        distance = compute_distance(self.node_id.id, target_id.id)

        # Находим первый ненулевой бит (слева направо)
        for i, byte in enumerate(distance):
            if byte != 0:
                # Находим позицию первого ненулевого бита в байте
                for bit in range(8):
                    if byte & (0x80 >> bit):
                        return i * 8 + bit

        # Если расстояние равно 0, это сам узел
        return len(self.buckets) - 1

    def add_node(self, node: Node) -> bool:
        """Добавление узла в таблицу маршрутизации"""
        if node.node_id == self.node_id:
            return False  # Не добавляем себя

        bucket_index = self._get_bucket_index(node.node_id)
        bucket = self.buckets[bucket_index]

        if bucket.is_full():
            # Если бакет полон, проверяем, есть ли устаревшие узлы
            stale_nodes = [n for n in bucket.nodes if n.is_stale()]
            if stale_nodes:
                # Заменяем устаревший узел
                bucket.remove_node(stale_nodes[0])
                return bucket.add_node(node)
            return False

        return bucket.add_node(node)

    def remove_node(self, node: Node):
        """Удаление узла из таблицы маршрутизации"""
        bucket_index = self._get_bucket_index(node.node_id)
        self.buckets[bucket_index].remove_node(node)

    def find_closest_nodes(self, target_id: NodeID, count: int) -> List[Node]:
        """Поиск ближайших узлов к целевому ID"""
        bucket_index = self._get_bucket_index(target_id)
        closest_nodes: List[Node] = []

        # Собираем узлы из текущего бакета и соседних
        for offset in range(len(self.buckets)):
            idx = (bucket_index + offset) % len(self.buckets)
            nodes = self.buckets[idx].get_nodes()
            closest_nodes.extend(nodes)

            if len(closest_nodes) >= count:
                break

        # Сортируем по расстоянию
        closest_nodes.sort(key=lambda n: compute_distance(n.node_id.id, target_id.id))

        return closest_nodes[:count]

    def get_all_nodes(self) -> List[Node]:
        """Получение всех узлов из таблицы"""
        all_nodes = []
        for bucket in self.buckets:
            all_nodes.extend(bucket.nodes)
        return all_nodes

File: src/dht/routing_table.rs
use crate::config::d_bucket_timeout;
use crate::dht::node::{Node, NodeID};
use std::time::{SystemTime, UNIX_EPOCH};

/// Return current time in seconds
fn get_now() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs_f64()
}

/// K-Buckets for saving nodes with their distance
pub struct KBucket {
    /// Volume of the bucket _(usually 20)_
    pub k: usize,
    /// List of nodes in this bucket
    pub nodes: Vec<Node>,
    /// Time of last update
    pub last_updated: f64,
}

impl KBucket {
    pub fn new(k: usize) -> Self {
        Self {
            k,
            nodes: Vec::with_capacity(k),
            last_updated: get_now(),
        }
    }

    /// Node add with LRU logic
    pub fn add_node(&mut self, node: Node) -> bool {
        if let Some(index) = self.nodes.iter().position(|n| n.node_id == node.node_id) {
            self.nodes.remove(index);
            self.nodes.push(node);
            self.last_updated = get_now();
            return true;
        }

        if self.nodes.len() < self.k {
            self.nodes.push(node);
            self.last_updated = get_now();
            return true;
        }

        false
    }

    /// Remove node from bucket
    pub fn remove_node(&mut self, node_id: &NodeID) {
        if let Some(index) = self.nodes.iter().position(|n| &n.node_id == node_id) {
            self.nodes.remove(index);
            self.last_updated = get_now();
        }
    }

    /// Get nodes from bucket
    pub fn get_nodes(&self) -> Vec<Node> {
        self.nodes.clone()
    }

    /// Check bucket state
    pub fn is_full(&self) -> bool {
        self.nodes.len() >= self.k
    }
}

/// Routing table
pub struct RoutingTable {
    /// Our node id
    pub node_id: NodeID,
    /// K volume of buckets
    pub k: usize,
    /// 160-counted buckets for 160-bits NodeId
    pub buckets: Vec<KBucket>,
}

impl RoutingTable {
    pub fn new(node_id: NodeID, k: usize, bucket_count: usize) -> Self {
        let mut buckets = Vec::with_capacity(bucket_count);
        for _ in 0..bucket_count {
            buckets.push(KBucket::new(k));
        }

        Self {
            node_id,
            k,
            buckets,
        }
    }

    /// Find bucket index for node id by XOR distance algo
    fn get_bucket_index(&self, target_id: &NodeID) -> usize {
        let distance = self.node_id.distance_to(target_id);

        for (i, &byte) in distance.iter().enumerate() {
            if byte != 0 {
                let leading_zeros = byte.leading_zeros() as usize;
                let index = i * 8 + leading_zeros;
                return index.min(self.buckets.len() - 1);
            }
        }

        self.buckets.len() - 1
    }

    /// Add node in routing table
    pub fn add_node(&mut self, node: Node) -> bool {
        if node.node_id == self.node_id {
            return false;
        }

        let bucket_index = self.get_bucket_index(&node.node_id);

        // Проверяем, полон ли бакет
        if self.buckets[bucket_index].is_full() {
            let stale_index = self.buckets[bucket_index]
                .nodes
                .iter()
                .position(|n| n.is_stale(d_bucket_timeout()));

            if let Some(idx) = stale_index {
                self.buckets[bucket_index].nodes.remove(idx);
                return self.buckets[bucket_index].add_node(node);
            }
            return false;
        }

        self.buckets[bucket_index].add_node(node)
    }

    /// Remove node
    pub fn remove_node(&mut self, node_id: &NodeID) {
        let bucket_index = self.get_bucket_index(node_id);
        self.buckets[bucket_index].remove_node(node_id);
    }

    /// Find closest nodes
    pub fn find_closest_nodes(&self, target_id: &NodeID, count: usize) -> Vec<Node> {
        let bucket_index = self.get_bucket_index(target_id);
        let mut closest_nodes: Vec<Node> = Vec::new();

        // Собираем узлы, начиная с целевого бакета и расширяя область
        for offset in 0..self.buckets.len() {
            let idx = (bucket_index + offset) % self.buckets.len();
            let nodes = self.buckets[idx].get_nodes();
            closest_nodes.extend(nodes);

            if closest_nodes.len() >= count * 2 {
                // Берем с запасом для точной сортировки
                break;
            }
        }

        // Сортируем по XOR расстоянию
        closest_nodes.sort_by_key(|n| n.node_id.distance_to(target_id));

        if closest_nodes.len() > count {
            closest_nodes.truncate(count);
        }

        closest_nodes
    }

    /// Getting all table nodes
    pub fn get_all_nodes(&self) -> Vec<Node> {
        self.buckets
            .iter()
            .flat_map(|bucket| bucket.nodes.clone())
            .collect()
    }
}
